services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/SwitchDeleteSpec.groovy

```diff
@@ -1,281 +1,194 @@
+        then: "Get 400 BadRequest error because the switch has ISLs"
         def exc = thrown(HttpClientErrorException)
         exc.rawStatusCode == 400
-        exc.responseBodyAsString.matches(".*Switch '$sw.dpId' has \\d+ assigned flows.*")
+        exc.responseBodyAsString.matches(".*Switch '$sw.dpId' has \\d+ $linkType links\\. $actionToFix.*")
 
-        and: "Cleanup: Activate the switch back and remove flow"
+        and: "Cleanup: activate the switch back"
         lockKeeper.reviveSwitch(sw.dpId)
-        Wrappers.wait(WAIT_OFFSET) {
-            northbound.activeSwitches.any { it.switchId == sw.dpId }
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.activeSwitches.any { it.switchId == sw.dpId } }
+
+        if (linkType == "inactive") {
+            swIsls.each { northbound.portUp(sw.dpId, it.srcPort) }
+            Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+                def links = northbound.getAllLinks()
+                swIsls.each { assert islUtils.getIslInfo(links, it).get().state == IslChangeType.DISCOVERED }
+            }
         }
-        northbound.deleteFlow(flow.id)
+
+        where:
+        [linkType, actionToFix] << [["active", "Unplug and remove them first"], ["inactive", "Remove them first"]]
     }
 
-    def "Cannot delete inactive switch with flow"() {
+    @Unroll
+    def "Unable to delete an inactive switch with a #flowType flow assigned"() {
         requireProfiles("virtual")
 
-        given: "An active switch with flow"
-        def (Switch srcSwitch, Switch dstSwitch) = topology.getActiveSwitches()[0..1]
-        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
-        northbound.addFlow(flow)
-        Wrappers.wait(WAIT_OFFSET) { northbound.getFlowStatus(flow.id).status == FlowState.UP }
+        given: "A flow going through a switch"
+        flowHelper.addFlow(flow)
 
-        when: "Deactivate this switch"
-        lockKeeper.knockoutSwitch(srcSwitch.getDpId())
-        Wrappers.wait(WAIT_OFFSET) {
-            !northbound.activeSwitches.any { it.switchId == srcSwitch.dpId}
-        }
+        when: "Deactivate the switch"
+        lockKeeper.knockoutSwitch(flow.source.datapath)
+        Wrappers.wait(WAIT_OFFSET) { assert !northbound.activeSwitches.any { it.switchId == flow.source.datapath } }
 
-        and: "Try to delete inactive switch with flow"
-        northbound.deleteSwitch(srcSwitch.getDpId(), false)
+        and: "Try to delete the switch"
+        northbound.deleteSwitch(flow.source.datapath, false)
 
-        then: "Got 400 BadRequest because switch has flow"
+        then: "Got 400 BadRequest error because the switch has the flow assigned"
         def exc = thrown(HttpClientErrorException)
         exc.rawStatusCode == 400
-        exc.responseBodyAsString.matches(".*Switch '$srcSwitch.dpId' has \\d+ assigned flows.*")
+        exc.responseBodyAsString.matches(".*Switch '${flow.source.datapath}' has \\d+ assigned flows.*")
```

Comment:
can we go the common way of deserializing the message into `MessageError` class? Also, the `\\d+` should be replaced with exact value that we expect here (which is `1`)

Reply:
Fixed.